mephisto/core/task_launcher.py

```diff
@@ -82,14 +96,57 @@ def create_assignments(self) -> None:
                     task_run.sandbox,
                 )
                 self.units.append(Unit(self.db, unit_id))
+                self.unlaunched_units[unit_id] = Unit(self.db, unit_id)
+
+    def generate_units(self):
+        """ units generator which checks that only 'max_num_concurrent_units' running at the same time,
+        i.e. in the LAUNCHED or ASSIGNED states """
+        while True:
```

Comment:
Generally it's good to avoid `while True` loops - here it'd be worthwhile to have a flag as to whether or not `expire_units` has been called, and exit this loop if it has.